File: R/ejercicio_02.R
enunciado <- "
=========================================================
EJERCICIO 2 (R)
De una baraja española de 40 cartas (4 palos: oros, copas, espadas, bastos; cada palo con números 1 al 7, sota, caballo, rey) se extraen dos cartas sin reemplazo.
a) ¿Cuál es la probabilidad de que ambas sean del mismo palo?
b) ¿Cuál es la probabilidad de que al menos una sea figura (sota, caballo o rey)?
=========================================================
"
cat(enunciado)

cat("--- SOLUCIÓN a) ---\n")
cat("Explicación: El total de formas de extraer 2 cartas de 40 sin orden es choose(40, 2).\n")
cat("Para que sean del mismo palo, elegimos 2 cartas de los 10 disponibles en un solo palo: choose(10, 2). Al haber 4 palos, multiplicamos esto por 4.\n")
total_extracciones <- choose(40, 2)
mismo_palo <- 4 * choose(10, 2)
prob_mismo_palo <- mismo_palo / total_extracciones
cat(sprintf("-> Casos totales (40C2): %d\n", total_extracciones))
cat(sprintf("-> Casos favorables (4 * 10C2): %d\n", mismo_palo))
cat(sprintf("-> Probabilidad = %.4f\n\n", prob_mismo_palo))

cat("--- SOLUCIÓN b) ---\n")
cat("Explicación: Calculamos la probabilidad de que NINGUNA sea figura y luego calculamos el complemento (1 - Probabilidad).\n")
cat("Hay 12 figuras (3 por palo) y 28 no-figuras. Las formas de sacar 2 no-figuras de las 28 es choose(28, 2).\n")
ninguna_figura <- choose(28, 2)
prob_ninguna <- ninguna_figura / total_extracciones
prob_al_menos_una <- 1 - prob_ninguna
cat(sprintf("-> Casos donde ninguna es figura (28C2): %d\n", ninguna_figura))
cat(sprintf("-> P(Ninguna figura) = %.4f\n", prob_ninguna))
cat(sprintf("-> P(Al menos una figura) = 1 - %.4f = %.4f\n", prob_ninguna, prob_al_menos_una))

fa <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
if (length(fa)) {
  source(file.path(dirname(normalizePath(sub("^--file=", "", fa[1]))), "utils_graficas.R"))
  carpeta <- carpeta_graficas(2)
  cargar_ggplot()
  df <- data.frame(
    evento = c("Mismo palo", "Al menos 1 figura"),
    prob = c(prob_mismo_palo, prob_al_menos_una)
  )
  p <- ggplot(df, aes(x = evento, y = prob, fill = evento)) +
    geom_col(width = 0.6, show.legend = FALSE) +
    scale_fill_manual(values = c("#9b59b6", "#e67e22")) +
    scale_y_continuous(limits = c(0, 1)) +
    labs(title = "Probabilidades del ejercicio 2", x = NULL, y = "Probabilidad") +
    tema_probabilidad()
  guardar_ggplot(p, carpeta, "probabilidades_cartas")
}

File: Python/ejercicio_02.py
import math

import matplotlib.pyplot as plt

from graficas_util import carpeta_graficas, guardar_figura


def crear_graficas(rojas, azules, verdes, espacio_muestral, formas_mismo_color):
    out = carpeta_graficas(2)
    colores = ["Rojas", "Azules", "Verdes"]
    casos = [math.comb(rojas, 2), math.comb(azules, 2), math.comb(verdes, 2)]
    plt.bar(colores, casos, color=["#e74c3c", "#3498db", "#2ecc71"])
    plt.ylabel("Formas de sacar 2 del mismo color")
    plt.title("Casos favorables por color (2 bolas)")
    guardar_figura(out, "mismo_color_por_tipo")

    plt.figure()
    plt.bar(
        ["Mismo color", "Distinto color"],
        [formas_mismo_color, espacio_muestral - formas_mismo_color],
        color=["#9b59b6", "#95a5a6"],
    )
    plt.ylabel("Número de combinaciones")
    plt.title(f"Espacio muestral (total = {espacio_muestral})")
    guardar_figura(out, "espacio_muestral")


def main():
    enunciado = """
=========================================================
EJERCICIO 2
En una urna hay 5 bolas rojas, 3 azules y 2 verdes. 
Se extraen dos bolas sin reemplazo.
a) ¿Cuántos elementos tiene el espacio muestral?
b) Calcule la probabilidad de que ambas bolas sean del mismo color.
=========================================================
"""
    print(enunciado)

    rojas = 5
    azules = 3
    verdes = 2
    total_bolas = rojas + azules + verdes

    print("--- SOLUCIÓN a) ---")
    print("Explicación: Se extraen 2 bolas de un total de 10 sin reemplazo. Como no importa el orden, usamos la fórmula de combinaciones: C(10, 2).")
    espacio_muestral = math.comb(total_bolas, 2)
    print(f"-> Elementos del espacio muestral (10C2): {espacio_muestral}\n")

    print("--- SOLUCIÓN b) ---")
    print("Explicación: Para que ambas sean del mismo color, sumamos las probabilidades de sacar 2 rojas O 2 azules O 2 verdes.")
    print("Calculamos las combinaciones posibles de extraer 2 bolas de cada color respectivo y las dividimos por el total del espacio muestral.")
    formas_rojas = math.comb(rojas, 2)
    formas_azules = math.comb(azules, 2)
    formas_verdes = math.comb(verdes, 2)
    
    formas_mismo_color = formas_rojas + formas_azules + formas_verdes
    prob_mismo_color = formas_mismo_color / espacio_muestral
    
    print(f"-> Formas de sacar 2 rojas (5C2): {formas_rojas}")
    print(f"-> Formas de sacar 2 azules (3C2): {formas_azules}")
    print(f"-> Formas de sacar 2 verdes (2C2): {formas_verdes}")
    print(f"-> Total de casos favorables: {formas_mismo_color}")
    print(f"-> Probabilidad = {formas_mismo_color} / {espacio_muestral} = {prob_mismo_color:.4f}")

    crear_graficas(rojas, azules, verdes, espacio_muestral, formas_mismo_color)


if __name__ == "__main__":
    main()
